ClientCredentialsKeypairs/HttpAuthHandler.cs

```diff
@@ -18,12 +21,48 @@ protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage
         {
             if (request.Options.All(x => x.Key != AnonymousOptionKey))
             {
-                var token = await _authTokenStore.GetToken();
-                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
+                var token = await _authTokenStore.GetToken(request.Method, request.RequestUri?.AbsoluteUri ?? "");
+                if (token != null)
+                {
+                    if (token.TokenType.ToUpper() == DpopSchemeType.ToUpper())
+                    {
+                        return await SendWithDpopAsync(request, cancellationToken, token);
+                    }
+                    else 
+                    {
+                        request.Headers.Authorization = new AuthenticationHeaderValue(token.TokenType, token.AccessToken);
+                    }
+                }
             }
 
             var response = await base.SendAsync(request, cancellationToken);
             return response;
         }
+
+        private async Task<HttpResponseMessage> SendWithDpopAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken, JwtAccessToken token)
+        {
+            request.Headers.Authorization = new AuthenticationHeaderValue(token.TokenType, token.AccessToken);
+
+            request.Headers.TryAddWithoutValidation(DpopHeaderName, token.DpopProof);
+
+            var dpopResponse = await base.SendAsync(request, cancellationToken);
+
+            if (dpopResponse.StatusCode == System.Net.HttpStatusCode.Unauthorized)
+            {
+                var supportedSchemes = dpopResponse.Headers.WwwAuthenticate.Select(x => x.Scheme).ToArray();
+
+                if (!supportedSchemes.Contains(DpopSchemeType, StringComparer.InvariantCultureIgnoreCase))
+                {
+                    // downgrade request to Dpop if Dpop is not supported
+                    request.Headers.Authorization = new AuthenticationHeaderValue(BearerSchemeType, token.AccessToken);
+                    request.Headers.Remove(DpopHeaderName);
+
+                    return await base.SendAsync(request, cancellationToken);
+                }
+            }
```

Comment:
Fallback support in case dpop is not supported on endpoint, should we allow this behavior considering we already have a feature flag? Ideally we could enable this functionality by a feature-flag as well.

Reply:
Fallback - øker jo kompleksiteten, syns egentlig det holder med featureflag, men at vi har bra feilmeldinger på dette.

Reply:
Men om han allerede har fungerende kode, så er det kanskje greit ?